---
title: "Dataset description"
author: "Philipp Janssen"
date: "24/08/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)
library(tidyseurat)

setwd("/data/share/htp/CZI_kidney/mouse_experiments/backgroundRNA/")
```

```{r color_schemes}
# cell types
celltype_colors <- setNames(
  RColorBrewer::brewer.pal(11, "Paired"),
  c("PCT","PECS-1/Thin_descending_limb","CD-A-IC/CD-B-IC","Mixed Immune","Mesangial","DCT","TAL","Endothelial","Renin_juxtaglomerular","Podocyte","CD-PC" ))

saveRDS(celltype_colors, "files/color_schemes/celltype_colors.RDS")

# strains
strain_colors = setNames(c("#D05957", "#4189C4", "#97B4DD"),
                         c("CAST_EiJ", "C57BL_6NJ", "129S1_SvImJ"))

strain_colors_simple = setNames(c("#D05957", "#4189C4", "#97B4DD"),
                         c("CAST", "BL6", "SvImJ"))

saveRDS(strain_colors, "files/color_schemes/strain_colors.RDS")

# methods 
method_colors <- setNames(c("#F8766D", "#7CAE00","#00BFC4"),
                          c("CellBender","DecontX","SoupX"))
saveRDS(method_colors, "files/color_schemes/method_colors.RDS")
```

## 1) Sample composition

### Prepare Seurat objects

**Input:** Initial Seurat objects (with celltype labels), strain annotations (from vireo)
**Aim:** Prepare Seurat objects in a uniform way to make them comparable between replicates. Rename cell type labels if necessary, remove cells without cell type or strain assignment, recalculate PCA & UMAP embeddings.
**Output:** filtered & processed Seurat objects (all cells & CAST cells only) 

```{r seurat, eval = FALSE}
# load Seurat objects & strain assignments
seurat_rep1 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment1_data/Seurat/seur.with.ann.RDS")
seurat_rep2 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment2_data/single_cell/Seurat/seur.ann.round2.RDS")
seurat_rep3 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment3_data/single_cell/Seurat/seur.RDS")
seurat_sn3 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment3_data/single_nucleus/Seurat/seur.with.celltypes.RDS")

# annotate Strains
donor_id_rep1 <- read_tsv("/data/share/htp/CZI_kidney/mouse_experiments/experiment1_data/cellSNP/supervised_allStrains.vireo.moreBC/donor_ids.tsv")
donor_id_rep2 <- read_tsv("/data/share/htp/CZI_kidney/mouse_experiments/experiment2_data/single_cell/cellSNP/supervised_allStrains.vireo/donor_ids.tsv")
donor_id_rep3 <- read_tsv("/data/share/htp/CZI_kidney/mouse_experiments/experiment3_data/single_cell/cellSNP/supervised_allStrains.vireo/donor_ids.tsv")
donor_id_sn3 <- read_tsv("/data/share/htp/CZI_kidney/mouse_experiments/experiment3_data/single_nucleus/cellSNP/supervised_allStrains.vireo/donor_ids.tsv")

add_strain_info <- function(seurat, donor_id){
  donor_id_mod <- donor_id %>% mutate(BC = gsub("-1", "", cell)) %>% dplyr::rename(Strain = donor_id) %>% dplyr::select(c(Strain, BC)) %>% column_to_rownames("BC") 
  seurat <- AddMetaData(seurat, donor_id_mod)
  print(paste0("removed ", sum(seurat$Strain %in% c("unassigned", "doublet")), " cells without strain assignment"))
  seurat <- seurat[,!(seurat$Strain %in% c("unassigned", "doublet"))]
  return(seurat)
}

seurat_rep1 <- add_strain_info(seurat_rep1, donor_id_rep1)
seurat_rep2 <- add_strain_info(seurat_rep2, donor_id_rep2)
seurat_rep3 <- add_strain_info(seurat_rep3, donor_id_rep3)
seurat_sn3 <- add_strain_info(seurat_sn3, donor_id_sn3)

# create uniform celltype labels
rename_celltypes <- function(seurat){
  seurat$CellType <- gsub("\\.", "-", seurat$CellType)
  #seurat$CellType <- gsub("\\?", "", seurat$CellType)
  seurat$CellType <- ifelse(seurat$CellType == "Podocytes", "Podocyte", seurat$CellType)
  seurat <- seurat[,seurat$CellType != "Other"]
  seurat <- seurat[,!(seurat$CellType %in% c("PCT?", "Mesangial?"))]
  return(seurat)
}

seurat_rep1 <- rename_celltypes(seurat_rep1)
seurat_rep2 <- rename_celltypes(seurat_rep2)
seurat_rep3 <- rename_celltypes(seurat_rep3)
seurat_sn3 <- rename_celltypes(seurat_sn3)

# pull out union of cell type for color scheme/ check if names are consistent across replicates
celltype_union <-  Reduce(union, list(unique(seurat_rep1$CellType), unique(seurat_rep2$CellType), unique(seurat_rep3$CellType), unique(seurat_sn3$CellType)))
celltype_union

# recalculate UMAP embedding after cell filtering (strain + cell type assignment)
run_UMAP <- function(seurat){
  seurat <- RunPCA(seurat)
  seurat <- RunUMAP(seurat, dims = 1:20)
}
seurat_rep1 <- run_UMAP(seurat_rep1)
seurat_rep2 <- run_UMAP(seurat_rep2)
seurat_rep3 <- run_UMAP(seurat_rep3)
seurat_sn3 <- run_UMAP(seurat_sn3)

saveRDS(seurat_rep1, "files/Seurat/seurat_rep1.RDS")
saveRDS(seurat_rep2, "files/Seurat/seurat_rep2.RDS")
saveRDS(seurat_rep3, "files/Seurat/seurat_rep3.RDS")
saveRDS(seurat_sn3, "files/Seurat/seurat_sn3.RDS")

# CAST cells only
seurat_rep1_CAST <- run_UMAP(filter(seurat_rep1, Strain == "CAST_EiJ"))
seurat_rep2_CAST <- run_UMAP(filter(seurat_rep2, Strain == "CAST_EiJ"))
seurat_rep3_CAST <- run_UMAP(filter(seurat_rep3, Strain == "CAST_EiJ"))
seurat_sn3_CAST <- run_UMAP(filter(seurat_sn3, Strain == "CAST_EiJ"))

saveRDS(seurat_rep1_CAST, "files/Seurat/seurat_rep1_CAST.RDS")
saveRDS(seurat_rep2_CAST, "files/Seurat/seurat_rep2_CAST.RDS")
saveRDS(seurat_rep3_CAST, "files/Seurat/seurat_rep3_CAST.RDS")
saveRDS(seurat_sn3_CAST, "files/Seurat/seurat_sn3_CAST.RDS")
```

### Plotting: UMAPs & composition (strain and cell types)

```{r plotting_functions_composition}
plot_UMAP <- function(seurat, color_by = "CellType", colors, title = NULL, plot_legend = F){
  p <- ggplot(seurat, aes_string(x = "UMAP_1", y = "UMAP_2", color = color_by))+
    geom_point(size = 0.1)+
    scale_color_manual(values = colors)+
    theme_bw()+
    theme(axis.title = element_blank(), axis.text = element_blank(), axis.ticks = element_blank())
  if(!is.null(title)){
    p <- p+ggtitle(title)
  }
  if(plot_legend == F){
    p <- p+theme(legend.position = "none")
  }
  
  return(p)
}

plot_strain_composition <- function(seurat, legend = F){
  p <- seurat %>% 
    mutate(Strain = ifelse(Strain == "129S5SvEvBrd", "C57BL_6NJ", Strain)) %>% 
    ggplot(aes(x = "a", fill = Strain))+
      geom_bar(position = "stack", stat = "count")+
      coord_flip()+
      scale_fill_manual(values = strain_colors)+
      labs(y = "number of cells")+
      theme_bw()+
      theme(axis.title = element_blank(), axis.text.y = element_blank(), legend.text.align = 0, axis.ticks.y = element_blank())#+
      #guides(fill = guide_legend(reverse = TRUE))
  
  ifelse(legend == T, return(p), return(p+theme(legend.position = "none")))
}

ct_freq <- seurat_rep1_CAST %>% 
  group_by(CellType) %>% 
  summarise(freq = n()) %>% 
  arrange(freq)
ct_order <- levels(reorder(ct_freq$CellType, ct_freq$freq))

plot_ct_composition <- function(seurat, legend = F, xlabel){
  p <- seurat %>% 
    mutate(CellType = factor(CellType, levels = ct_order)) %>% 
    ggplot(aes(x = xlabel, fill = CellType))+
      geom_bar(position = "stack")+
      coord_flip()+
      scale_fill_manual(values = celltype_colors, breaks = rev(ct_order))+
      #scale_y_continuous(limits = c(0,length(seurat$CellType)))+
      labs(y = "number of cells")+
      theme_bw()+
      theme(axis.title = element_blank(), 
            #axis.text.y = element_blank(), legend.text.align = 0, axis.ticks.y = element_blank()
            )
  
  ifelse(legend == T, return(p), return(p+theme(legend.position = "none")))
}
```

```{r UMAPs, fig.width=11, fig.height=8}
ct_legend <- get_legend(
  ggplot(seurat_rep1, aes_string(x = "UMAP_1", y = "UMAP_2", color = "CellType"))+
    geom_point(size = 4)+
    scale_color_manual(values = celltype_colors)+
    theme(legend.position = "bottom",legend.justification = c(0,1))+
    guides(color=guide_legend(nrow=2,byrow=TRUE))
)

strain_legend <- get_legend(
  ggplot(seurat_rep1, aes_string(x = "UMAP_1", y = "UMAP_2", color = "Strain"))+
    geom_point(size = 4)+
    scale_color_manual(values = strain_colors)+
    theme(legend.position = "bottom",legend.justification = c(0,1))+
    guides(color=guide_legend(nrow=1,byrow=TRUE))
)

p.UMAPs <- plot_grid(
  # All cells: cell types
  plot_grid(
    plot_UMAP(seurat_rep1, colors = celltype_colors, title = "single-cell - rep1"),
    plot_UMAP(seurat_rep2, colors = celltype_colors, title = "single-cell - rep2"),
    plot_UMAP(seurat_rep3, colors = celltype_colors, title = "single-cell - rep3"),
    plot_UMAP(seurat_sn3, colors = celltype_colors, title = "single-nucleus - rep3"), 
    #get_legend(plot_UMAP(seurat_rep1, colors = celltype_colors, plot_legend = T)),
    nrow = 1
  ),
  # All cells: strains
  plot_grid(
    plot_UMAP(seurat_rep1, color_by = "Strain", colors = strain_colors),
    plot_UMAP(seurat_rep2, color_by = "Strain", colors = strain_colors),
    plot_UMAP(seurat_rep3, color_by = "Strain", colors = strain_colors),
    plot_UMAP(seurat_sn3, color_by = "Strain", colors = strain_colors), 
    #get_legend(plot_UMAP(seurat_rep1, color_by = "Strain", colors = strain_colors, plot_legend = T)),
    nrow = 1
  ),
  # CAST cells: cell types
  plot_grid(
    plot_UMAP(seurat_rep1_CAST, colors = celltype_colors, title = "CAST cells only"),
    plot_UMAP(seurat_rep2_CAST, colors = celltype_colors, title = "CAST cells only"),
    plot_UMAP(seurat_rep3_CAST, colors = celltype_colors, title = "CAST cells only"),
    plot_UMAP(seurat_sn3_CAST, colors = celltype_colors, title = "CAST cells only"), 
    #get_legend(plot_UMAP(seurat_rep1_CAST, colors = celltype_colors, plot_legend = T)),
    nrow = 1
  ),
  
  plot_grid(
    ct_legend, strain_legend, nrow = 2, align = "v"
  ),
  
  nrow = 4, rel_heights = c(1.1,1,1.1,0.7)
)

p.UMAPs

ggsave("figures/01_dataset_description/UMAPs.pdf", p.UMAPs, width = 10, height = 8)
```

```{r composition}
p.ct_comp <-plot_grid(
  plot_ct_composition(seurat_rep1_CAST, xlabel = "rep1"),
  plot_ct_composition(seurat_rep2_CAST, xlabel = "rep2"),
  plot_ct_composition(seurat_rep3_CAST, xlabel = "rep3"),
  NULL,
  nrow = 4, labels = c("","","","number of cells"), label_size = 8, label_x = 0, label_y = 1.15,
  rel_heights = c(1,1,1,0.1))

p.strain_comp <- plot_grid(
  plot_strain_composition(seurat_rep1),
  plot_strain_composition(seurat_rep2),
  plot_strain_composition(seurat_rep3), 
 NULL,
  nrow = 4, labels = c("","","","number of cells"), label_size = 8, label_x = -0.1, label_y = 1.15,
 rel_heights = c(1,1,1,0.1))

plot_grid(
  p.ct_comp,
  p.strain_comp,
  nrow =1, rel_widths = c(1,0.9)
)

```

Combination of example UMAP and composition bar plots for final figure: 

```{r UMAP_and_composition}
ct_legend_comp <- get_legend(plot_ct_composition(seurat_rep1_CAST, xlabel = "rep1", legend = T)+
                               theme(legend.position = "bottom", legend.title = element_blank())+
                               guides(fill=guide_legend(nrow=3,byrow=F)))

strain_legend_comp <- get_legend(plot_strain_composition(seurat_rep1, legend = T)+
                               theme(legend.position = "bottom", legend.title = element_blank())+
                               guides(color=guide_legend(nrow=1,byrow=F)))

plot_grid(
  plot_grid(
    plot_UMAP(seurat_rep1_CAST, colors = celltype_colors),
    p.ct_comp, 
    p.strain_comp, 
    nrow = 1, 
    rel_widths = c(1.4,1,0.9)#, 
    #labels = c("B","C","D")
  ),
  plot_grid(
    ct_legend_comp,
    strain_legend_comp,
    nrow = 1, rel_widths = c(2.4,0.9)
  ),
  nrow = 2, rel_heights = c(1,0.3))
```

## 2) Genotype estimation statistics

```{r import_csc}
csc_rep1 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment1_data/GT_based_background_RNA/new/moreBC/csc.RDS")
csc_rep2 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment2_data/single_cell/background_RNA/GT_based/new/csc.RDS")
csc_rep3 <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment3_data/single_cell/background_RNA/GT_based/csc.RDS")

csc_rep1_noMito <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment1_data/GT_based_background_RNA/new/moreBC/csc_noMito.RDS")
csc_rep2_noMito <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment2_data/single_cell/background_RNA/GT_based/new/csc_noMito.RDS")
csc_rep3_noMito <- readRDS("/data/share/htp/CZI_kidney/mouse_experiments/experiment3_data/single_cell/background_RNA/GT_based/csc_noMito.RDS")
```

### number of positions and genes with coverage for strain specific SNPs

```{r nPos}
position_stats <- rbind(
  csc_rep1 %>% select(position, ENSEMBL, SNPisALT) %>% distinct,
  csc_rep2 %>% select(position, ENSEMBL, SNPisALT) %>% distinct,
  csc_rep3 %>% select(position, ENSEMBL, SNPisALT) %>% distinct
) %>% 
  distinct()

position_stats_noMito <- rbind(
  csc_rep1_noMito %>% select(position, ENSEMBL, SNPisALT) %>% distinct,
  csc_rep2_noMito %>% select(position, ENSEMBL, SNPisALT) %>% distinct,
  csc_rep3_noMito %>% select(position, ENSEMBL, SNPisALT) %>% distinct
) %>% 
  distinct()

plot_nPos_nGene <- function(position_stats){
  position_stats_summary <- position_stats %>% 
    mutate(strain_specific = ifelse(SNPisALT == "CAST_SvImJ", "BL6", SNPisALT)) %>% 
    group_by(strain_specific) %>% 
    summarize(nPos = length(unique(position)),
              nGene = length(unique(ENSEMBL))) %>% 
    arrange(nPos) %>% 
    mutate(lab.ypos = cumsum(nPos) - 0.5*nPos,
           lab.ypos.gene = cumsum(nGene) - 0.5*nGene,
           strain_specific = factor(strain_specific, levels = c("CAST","SvImJ","BL6")))
  
  p.pie_nPos <- ggplot(position_stats_summary, aes(x = "", y = nPos, fill = strain_specific))+
    geom_bar(stat = "identity", width = 1, color = "white")+
    coord_polar("y", start = 0)+
    geom_text(aes(y = lab.ypos, label = nPos), color = "white")+
    scale_fill_manual(values = strain_colors_simple) +
    theme_void()+
    ggtitle("Number of strain-specific SNPs")+
    theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5))
  
  p.bar_nGene <- ggplot(position_stats_summary, aes(x = strain_specific, y = nGene, fill = strain_specific))+
    geom_bar(stat = "identity", width = 1, color = "white")+
    coord_flip()+
    #coord_polar("y", start = 0)+
    geom_text(aes(y = 0.5*nGene, label = nGene), color = "white")+
    scale_fill_manual(values = strain_colors_simple) +
    theme_bw()+
    labs(y = "genes with strain-specific SNPs")+
    theme(legend.position = "none",  axis.title.y = element_blank())
  
  plot_grid(
    p.pie_nPos,
    p.bar_nGene,
    nrow = 2, rel_heights = c(1,0.5)
  )
}

plot_nPos_nGene(position_stats)
plot_nPos_nGene(position_stats_noMito)

ggsave("figures/01_dataset_description/strain_specific_SNPs.pdf", plot_nPos_nGene(position_stats_noMito), width = 5, height = 5)
```

### number of SNPs/genes per cell

Looking at CAST specific SNPs in CAST cells only:

```{r nPos_nGene_per_cell}
per_cell_stats <- rbind(
    csc_rep1_noMito %>% 
  filter(SNPisALT == "CAST", Strain == "CAST", cell %in% paste0(colnames(seurat_rep1_CAST),"-1")) %>% 
  group_by(cell) %>% 
  summarize(sum_DP = sum(DP), nGene = length(unique(ENSEMBL)), nPos = length(unique(position))) %>% 
    mutate(replicate = "rep1"),
  
    csc_rep2_noMito %>% 
  filter(SNPisALT == "CAST", Strain == "CAST", cell %in% paste0(colnames(seurat_rep2_CAST),"-1")) %>% 
  group_by(cell) %>% 
  summarize(sum_DP = sum(DP), nGene = length(unique(ENSEMBL)), nPos = length(unique(position))) %>% 
    mutate(replicate = "rep2"),
    
  csc_rep3_noMito %>% 
  filter(SNPisALT == "CAST", Strain == "CAST", cell %in% paste0(colnames(seurat_rep3_CAST),"-1")) %>% 
  group_by(cell) %>% 
  summarize(sum_DP = sum(DP), nGene = length(unique(ENSEMBL)), nPos = length(unique(position))) %>% 
    mutate(replicate = "rep3")
) %>% 
  mutate(replicate = factor(replicate, levels = c("rep3","rep2","rep1")))


p.SNP_coverage_perCell <- plot_grid(
  ggplot(per_cell_stats, aes(x = replicate, y = sum_DP))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank()),
  
  ggplot(per_cell_stats, aes(x = replicate, y = nGene))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "genes with UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank(), axis.text.y = element_blank()), 

  nrow = 1, align = "hv"
)  

p.SNP_coverage_perCell

ggsave("figures/01_dataset_description/SNP_coverage_per_Cell.pdf", p.SNP_coverage_perCell, width = 6, height = 2.5)


plot_grid(
  ggplot(per_cell_stats, aes(x = replicate, y = sum_DP))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank()),
  
  ggplot(per_cell_stats, aes(x = replicate, y = nPos))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "SNPs with UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank()),
  
  ggplot(per_cell_stats, aes(x = replicate, y = nGene))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "genes with UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank(), axis.text.y = element_blank()), 

  nrow = 1, align = "hv"
)  
```
