---
title: "Dataset description"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/data/share/htp/CZI_kidney/mouse_experiments/backgroundRNA")

library(tidyverse)
library(tidyseurat)
library(cowplot)
```


```{r color_schemes}
strain_colors = setNames(c("#D05957", "#4189C4", "#97B4DD", "lightgrey"),
                         c("CAST", "BL6", "SvImJ", "doublet"))

denisenko_colors <- setNames(
  c(RColorBrewer::brewer.pal(11, "Paired"),"darkgrey","grey"),
  c("PT","CD_IC" ,   "CD_PC", "CD_Trans" ,     "CNT" ,     "DCT",     "Endo",      "Fib","aLOH",   "dLOH",    "MC" ,    "Podo","Immune"))

```

## Figure 1

```{r load_data}
cell_metadata <- readRDS("Extra_analyses/cell_metadata.RDS")

cell_metadata <- cell_metadata %>% 
   filter(Strain != "unassigned") %>% 
   mutate(Strain = factor(Strain, levels = rev(c("CAST","BL6","SvImJ","doublet"))),
          replicate = factor(replicate, levels = c("rep1","rep2","rep3","nuc2","nuc3")))

```


### A) experimental design scheme

```{r experimental_design}
exp_design <- ggdraw() + draw_image("figures/experimental_design.png")
#exp_design <- ggdraw() + draw_image( magick::image_read_svg("figures/Figure1A_schematic_overview.svg"))
```

### B) strain composition per replicate

```{r figure_1b, fig.width=5, fig.height=4}
p.strain_composition <- cell_metadata %>% 
  ggplot(aes(x = "a", fill = Strain))+
        geom_bar(position = "stack", stat = "count")+
        coord_flip()+
        scale_fill_manual(values = strain_colors)+
        facet_grid(replicate~., switch = "y")+
        labs(y = "number of cells")+
        theme_bw()+
        theme(axis.title.y = element_blank(), axis.text.y = element_blank(), legend.text.align = 0, axis.ticks.y = element_blank(),
              legend.position = "bottom", legend.title = element_blank())

p.strain_composition
```


### C) cell type composition per replicate

```{r figure_1c, fig.width=5, fig.height=4}
p.celltype_composition <- cell_metadata %>% 
  filter(Strain != "doublet", !is.na(celltype)) %>% 
  mutate(celltype = factor(celltype, levels = names(sort(table(.$celltype))))) %>% 
  ggplot(aes(x = Strain, fill = celltype))+
        geom_bar(position = "fill")+
        coord_flip()+
        scale_fill_manual(values =denisenko_colors)+
        facet_grid(replicate~.)+
        labs(y = "fraction of cells")+
        theme_bw()+
        theme(axis.title.y = element_blank(),legend.position = "bottom", legend.title = element_blank(),strip.text.y.right = element_text(angle = 90))+
        guides(fill=guide_legend(nrow=2))

p.celltype_composition
```


Strain and cell type fractions:
```{r strain_fractions}
cell_metadata %>% 
  filter(Strain != "doublet") %>% 
  group_by(replicate, Strain) %>% 
  dplyr::summarize(n = n()) %>% 
  group_by(replicate) %>% 
  mutate(frac = n / sum(n))

cell_metadata %>% 
  filter(Strain != "doublet") %>% 
  group_by(replicate, celltype) %>% 
  dplyr::summarize(n = n()) %>% 
  group_by(replicate) %>% 
  mutate(frac = n / sum(n)) %>% View
```


### D) UMAP of replicate 2

```{r figure_1d, fig.width=5, fig.height=4}
seurat_rep2 <- readRDS("Snakemake_backup/input/rep2/seurat_CAST.RDS")

p.UMAP <- ggplot(seurat_rep2, aes(x = UMAP_1, y = UMAP_2, color = celltype))+
    geom_point(size = 0.5)+
    scale_color_manual(values = denisenko_colors)+
    theme_bw()+
    theme(axis.text = element_blank(), axis.ticks = element_blank(), legend.position = "none")

p.UMAP
```


### E) analysis workflow scheme

```{r figure_1e}
analysis_scheme <- ggdraw() + draw_image("figures/Figure1E_analysis_scheme.png")
```

### Combine plots & assemble figure

```{r figure1_cde, fig.height=9, fig.width=10}
p.figure1 <- plot_grid(
  # A: experimental design
  exp_design,
  
  NULL,
  
  # B,C,D: data composition
  plot_grid(
    plot_grid(p.strain_composition+theme(legend.position = "none"), 
              p.celltype_composition+theme(legend.position = "none"), 
              p.UMAP, 
              nrow = 1, align = "h", scale = c(0.9,0.9,0.86), 
              labels = c("B","C","D")),
    plot_grid(get_legend(p.strain_composition), get_legend(p.celltype_composition), nrow = 1, rel_widths = c(1,2)),
    nrow = 2, rel_heights = c(1,0.15)
  ),
  
  NULL,
  
  # E: schematic analysis overview
  analysis_scheme,
  
  nrow = 5, labels = c("A","","","","E"), rel_heights = c(0.7,0.05,1,0.1,0.9)
)

p.figure1

ggsave("figures/Figure1.pdf", p.figure1, width = 10, height = 9)
```



## Supplementary figure 1

```{r suppl_fig1, fig.width=14, fig.height=8}
# load seurat objects
seu_list <- list()
seu_list_CAST <- list()

for(exp in c("rep1", "rep2", "rep3", "nuc2", "nuc3")){
    seu_list[[exp]] <- readRDS(paste0("Snakemake_backup/input/",exp,"/seurat.RDS"))
    seu_list_CAST[[exp]] <- readRDS(paste0("Snakemake_backup/input/",exp,"/seurat_CAST.RDS"))
}


plot_UMAP <- function(seurat, color_by = "celltype", colors, title = NULL, plot_legend = F){
  p <- ggplot(seurat, aes_string(x = "UMAP_1", y = "UMAP_2", color = color_by))+
    geom_point(size = 0.1)+
    scale_color_manual(values = colors)+
    theme_bw()+
    theme(axis.title = element_blank(), axis.text = element_blank(), axis.ticks = element_blank())
  if(!is.null(title)){
    p <- p+ggtitle(title)
  }
  if(plot_legend == F){
    p <- p+theme(legend.position = "none")
  }
  
  return(p)
}

p.UMAPs <- plot_grid(
  # All cells: strains
  plot_grid(
    plot_UMAP(seu_list$rep1, color_by = "Strain", title = "single-cell - rep1",colors = strain_colors)+theme(title = element_text(size = 14)),
    plot_UMAP(seu_list$rep2, color_by = "Strain", title = "single-cell - rep2",colors = strain_colors)+theme(title = element_text(size = 14)),
    plot_UMAP(seu_list$rep3, color_by = "Strain", title = "single-cell - rep3",colors = strain_colors)+theme(title = element_text(size = 14)),
    plot_UMAP(seu_list$nuc2, color_by = "Strain", title = "single-nucleus - rep2",colors = strain_colors)+theme(title = element_text(size = 14)),
    get_legend(plot_UMAP(seu_list$rep1, color_by = "Strain" ,colors = strain_colors, plot_legend = T)+guides(colour = guide_legend(override.aes = list(size=5)))+theme(legend.position = c(0.2,0.5))),
    nrow = 1, scale = c(0.92,0.92,0.92,0.92)
  ),
  # All cells: cell types
  plot_grid(
    plot_UMAP(seu_list$rep1, colors = denisenko_colors),
    plot_UMAP(seu_list$rep2, colors = denisenko_colors),
    plot_UMAP(seu_list$rep3, colors = denisenko_colors),
    plot_UMAP(seu_list$nuc2, colors = denisenko_colors), 
    get_legend(plot_UMAP(seu_list$rep1, colors = denisenko_colors,  plot_legend = T)+guides(colour = guide_legend(override.aes = list(size=5)))+theme(legend.position = c(0.25,0))),
    nrow = 1,  scale = c(0.92,0.92,0.92,0.92)
  ),
  # CAST cells: cell types
  plot_grid(
    plot_UMAP(seu_list_CAST$rep1, colors = denisenko_colors, title = "CAST cells only"),
    plot_UMAP(seu_list_CAST$rep2, colors = denisenko_colors, title = "CAST cells only"),
    plot_UMAP(seu_list_CAST$rep3, colors = denisenko_colors, title = "CAST cells only"),
    plot_UMAP(seu_list_CAST$nuc2, colors = denisenko_colors, title = "CAST cells only"),
    NULL,
    #get_legend(plot_UMAP(seu_list$rep1_CAST, colors = denisenko_colors, plot_legend = T)),
    nrow = 1,  scale = c(0.92,0.92,0.92,0.92)
  ),
  
  #plot_grid(
  #  ct_legend, strain_legend, nrow = 2, align = "v"
  #),
  
  nrow = 3, rel_heights = c(1.1,1,1.1), labels = c("A","B","C")
)



p.UMAPs <- plot_grid(
  # All cells: strains
  plot_grid(
    plot_UMAP(seu_list$rep1, color_by = "Strain", title = "single-cell - rep1",colors = strain_colors)+theme(title = element_text(size = 12)),
    plot_UMAP(seu_list$rep2, color_by = "Strain", title = "single-cell - rep2",colors = strain_colors)+theme(title = element_text(size = 12)),
    plot_UMAP(seu_list$rep3, color_by = "Strain", title = "single-cell - rep3",colors = strain_colors)+theme(title = element_text(size = 12)),
    plot_UMAP(seu_list$nuc2, color_by = "Strain", title = "single-nucleus - nuc2",colors = strain_colors)+theme(title = element_text(size = 12)),
    plot_UMAP(seu_list$nuc3, color_by = "Strain", title = "single-nucleus - nuc3",colors = strain_colors)+theme(title = element_text(size = 12)),
    get_legend(plot_UMAP(seu_list$rep1, color_by = "Strain" ,colors = strain_colors, plot_legend = T)+guides(colour = guide_legend(override.aes = list(size=5)))+theme(legend.position = c(0.2,0.5))),
    nrow = 1, scale = c(0.92,0.92,0.92,0.92,0.92)
  ),
  # All cells: cell types
  plot_grid(
    plot_UMAP(seu_list$rep1, colors = denisenko_colors),
    plot_UMAP(seu_list$rep2, colors = denisenko_colors),
    plot_UMAP(seu_list$rep3, colors = denisenko_colors),
    plot_UMAP(seu_list$nuc2, colors = denisenko_colors), 
    plot_UMAP(seu_list$nuc3, colors = denisenko_colors), 
    get_legend(plot_UMAP(seu_list$rep1, colors = denisenko_colors,  plot_legend = T)+guides(colour = guide_legend(override.aes = list(size=5)))+theme(legend.position = c(0.25,0))),
    nrow = 1,  scale = c(0.92,0.92,0.92,0.92,0.92)
  ),
  # CAST cells: cell types
  plot_grid(
    plot_UMAP(seu_list_CAST$rep1, colors = denisenko_colors, title = "CAST cells only")+theme(title = element_text(size = 10)),
    plot_UMAP(seu_list_CAST$rep2, colors = denisenko_colors, title = "CAST cells only")+theme(title = element_text(size = 10)),
    plot_UMAP(seu_list_CAST$rep3, colors = denisenko_colors, title = "CAST cells only")+theme(title = element_text(size = 10)),
    plot_UMAP(seu_list_CAST$nuc2, colors = denisenko_colors, title = "CAST cells only")+theme(title = element_text(size = 10)),
    plot_UMAP(seu_list_CAST$nuc3, colors = denisenko_colors, title = "CAST cells only")+theme(title = element_text(size = 10)),
    NULL,
    #get_legend(plot_UMAP(seu_list$rep1_CAST, colors = denisenko_colors, plot_legend = T)),
    nrow = 1,  scale = c(0.92,0.92,0.92,0.92,0.92)
  ),
  
  #plot_grid(
  #  ct_legend, strain_legend, nrow = 2, align = "v"
  #),
  
  nrow = 3, rel_heights = c(1.1,1,1.1), labels = c("A","B","C")
)

p.UMAPs


ggsave("figures/Suppl_figure1.pdf", p.UMAPs, width = 14, height = 8)

```

## Supplementary figure 2

Summarize stats for genotype estimation strategy:

```{r summarize_stats, fig.width=10, fig.height=3.4}
csc_list <- list()

for(exp in c("rep1", "rep2", "rep3", "nuc2", "nuc3")){
    csc_list[[exp]] <- readRDS(paste0("Snakemake_backup/genotype_estimation/",exp,"/csc_intermediate.RDS"))
}

csc_full <- bind_rows(csc_list, .id = "replicate")

# Filtering: remove variants in mitochondrial genes, intergenic variants & empty droplets/bad cells
csc_full <- filter(csc_full, 
                   !grepl("mt-", SYMBOL), 
                   !is.na(ENSEMBL),
                   cell %in% cell_metadata$cell)


position_stats_summary <- csc_full %>% 
    mutate(strain_specific = ifelse(SNPisALT == "CAST_SvImJ", "BL6", SNPisALT)) %>% 
    group_by(strain_specific) %>% 
    summarize(nPos = length(unique(position)),
              nGene = length(unique(ENSEMBL))) %>% 
    arrange(nPos) %>% 
    mutate(lab.ypos = cumsum(nPos) - 0.5*nPos,
           lab.ypos.gene = cumsum(nGene) - 0.5*nGene,
           strain_specific = factor(strain_specific, levels = c("CAST","SvImJ","BL6")))

saveRDS(position_stats_summary, "Extra_analyses/position_stats_summary.RDS")


p.pie_nPos <- ggplot(position_stats_summary, aes(x = "", y = nPos, fill = strain_specific))+
  geom_bar(stat = "identity", width = 1, color = "white")+
  coord_polar("y", start = 0)+
  geom_text(aes(y = lab.ypos, label = nPos), color = "white")+
  scale_fill_manual(values = strain_colors, limits = force) +
  theme_void()+
  labs(caption = "Number of strain-specific SNPs")+
  theme(legend.title = element_blank(),# plot.title = element_text(hjust = 0.5),
        plot.caption = element_text(hjust=0.5, size=rel(1), vjust = 5))

p.bar_nGene <- ggplot(position_stats_summary, aes(x = strain_specific, y = nGene, fill = strain_specific))+
  geom_bar(stat = "identity", width = 1, color = "white")+
  coord_flip()+
  #coord_polar("y", start = 0)+
  geom_text(aes(y = 0.5*nGene, label = nGene), color = "white")+
  scale_fill_manual(values = strain_colors) +
  theme_bw()+
  labs(y = "genes with strain-specific SNPs")+
  theme(legend.position = "none",  axis.title.y = element_blank())

p.position_stats <- plot_grid(
  p.pie_nPos,
  p.bar_nGene,
  nrow = 2, rel_heights = c(1,0.5), labels = c("A", "B"))
  
```  
  
```{r per_cell_stats}
# per cell stats for CAST-specific SNPs only
  
per_cell_stats <- csc_full %>% 
  filter(SNPisALT == "CAST",
         Strain %in% c("CAST", "BL6", "SvImJ")) %>% 
  group_by(replicate, Strain, cell) %>% 
  summarize(sum_DP = sum(DP), nGene = length(unique(ENSEMBL)), nPos = length(unique(position)))

saveRDS(per_cell_stats, "Extra_analyses/per_cell_stats_CAST_variants.RDS")

per_cell_stats_CAST <-  filter(per_cell_stats, Strain == "CAST") %>% 
  mutate(replicate = factor(replicate, levels = c("nuc3","nuc2", "rep3", "rep2", "rep1")))

p.SNP_coverage_perCell <- plot_grid(
  ggplot(per_cell_stats, aes(x = replicate, y = sum_DP, fill = Strain))+
    #geom_violin()+
    geom_boxplot(width = 0.8, outlier.colour = NA)+
    scale_fill_manual(values = strain_colors)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank()),
  
  ggplot(per_cell_stats, aes(x = replicate, y = nGene, fill = Strain))+
    #geom_violin()+
    geom_boxplot(width = 0.8, outlier.colour = NA)+
     scale_fill_manual(values = strain_colors)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "genes with UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank(), axis.text.y = element_blank()), 

  nrow = 1, align = "hv"
)  

p.SNP_coverage_perCell


# CAST cells only: 

p.SNP_coverage_perCell_CAST <- plot_grid(
  ggplot(per_cell_stats_CAST, aes(x = replicate, y = sum_DP))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank()),
  
  ggplot(per_cell_stats_CAST, aes(x = replicate, y = nGene))+
    geom_violin()+
    geom_boxplot(width = 0.2, outlier.colour = NA)+
    scale_y_log10()+
    coord_flip()+
    labs(y = "genes with UMI coverage per cell")+
    theme_bw()+
    theme(axis.title.y = element_blank(), axis.text.y = element_blank()), 

  nrow = 1, align = "hv", labels = c("C", "D")
)  

p.SNP_coverage_perCell_CAST


p.suppl2 <- plot_grid(p.position_stats, p.SNP_coverage_perCell_CAST, rel_widths = c(1,1.7))

p.suppl2

ggsave("figures/Suppl_figure2.pdf", p.suppl2, width = 10, height = 3.4)

# per CAST cell, modify plot:
p.per_cell_coverage <- per_cell_stats_CAST %>% 
  mutate(replicate = factor(replicate, levels = c("rep1","rep2","rep3", "nuc2","nuc3"))) %>% 
   ggplot(aes(x = replicate, y = sum_DP))+
      geom_violin(aes(fill = replicate), scale = "width")+
      geom_boxplot(width = 0.2, outlier.colour = NA)+
      scale_fill_manual(values = replicate_colors)+
      scale_y_log10()+
      #coord_flip()+
      labs(y = "Coverage of informative variants (UMIs/cell)")+
      theme_bw()+
      theme(axis.title.x = element_blank(),
            legend.position = "none",
            axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


### Per Gene: 

per_gene_stats <- csc_full %>% 
  group_by(replicate, Strain, SYMBOL) %>% 
  summarize(sum_DP = sum(DP), nCell = length(unique(cell)), nPos = length(unique(position)))

dim(per_gene_stats)
View(per_gene_stats)

```


## Poster figures


Cell type composition: 

```{r poster_plots, fig.width=8, fig.height=4}
p.ct_comp <- plot_grid(
    plot_grid(p.celltype_composition+theme(legend.position = "none"), 
              p.UMAP, 
              nrow = 1, align = "hv", scale = c(0.9,0.9)), 
    get_legend(p.celltype_composition),
    nrow = 2, rel_heights = c(1,0.15)
)

ggsave("figures/SCOG_poster/ct_composition.pdf", p.ct_comp, width = 8, height = 4)

p.ct_comp
```

Strain composition & informative variants: 

```{r poster_plots, fig.width=9, fig.height=3.7}
p.variants <- plot_grid(
    plot_grid(p.strain_composition+theme(legend.position = "none"), 
              p.pie_nPos+theme(legend.position = "none"), 
              p.per_cell_coverage,
              nrow = 1, align = "hv", scale = c(0.95,0.95,0.95), rel_widths = c(1,1,0.74)), 
    plot_grid(get_legend(p.strain_composition), NULL, rel_widths = c(2,0.74)),
    nrow = 2, rel_heights = c(1,0.15)
)


p.variants

ggsave("figures/SCOG_poster/strain_composition.pdf", p.variants, width = 9, height = 3.7)
```


## Updated final figure 1

```{r figure1_assembly, fig.width=10, fig.height=10}
p.figure1 <- plot_grid(
  # Experimental design scheme
  exp_design, 
  
  # Strain composition and informative variant stats
  plot_grid(
    plot_grid(p.strain_composition+theme(legend.position = "none"), 
              p.pie_nPos+theme(legend.position = "none"), 
              p.per_cell_coverage,
              nrow = 1, align = "hv", scale = c(0.95,0.95,0.85), rel_widths = c(0.8,1,0.8), labels = c("B","C","D")), 
    plot_grid(get_legend(p.strain_composition), NULL, rel_widths = c(2,0.74)),
    nrow = 2, rel_heights = c(1,0.15)),
  
  # Cell type composition and example UMAP
  plot_grid(
    plot_grid(p.celltype_composition+theme(legend.position = "none"), 
              p.UMAP, 
              nrow = 1, align = "hv", scale = c(0.9,0.9), labels = c("E", "F")), 
    get_legend(p.celltype_composition),
    nrow = 2, rel_heights = c(1,0.15)),
  
  nrow = 3, labels = c("A","",""), rel_heights = c(0.8,1,1.2)
  )

p.figure1

ggsave("figures/Figure1.pdf", p.figure1, width = 10, height = 10)
```
